### ISSUE TITLE
Scattering error

### ISSUE THREAD

**Issue by mckinsel** [dnanexus/dxWDL#146]
I'm trying to run the pipeline defined [here](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/optimus/Optimus.wdl) with dxWDL. It's failing in the `scatter (index in indices)` job with the error `A job in the job tree named "scatter (index in indices)" running function "main" of "Optimus_scatter_index" failed because of AppInternalError: Unsupported combination: womType: File womValue: ([ dx: file-FGqBJbj05pkb4kfG19YbXJz3:: 78c92f51-76b6-4f98-8e0f-c00f4b732e42 pbmc8k_S1_L007_R1_001.fastq.gz , dx: file-FGqBJbj05pkX3vzF8PPx7kZB:: 78c92f51-76b6-4f98-8e0f-c00f4b732e42 pbmc8k_S1_L007_R1_001.fastq.gz ], Array[File])`

The generated workflow code is below. Note that in the workflow definition has `r1_fastq` as an `Array[File]` but in the task, it's a `File`. Any thoughts?

```
Workflow source code:
task Optimus__scatter_index_body {
    Array[File] r2_fastq 
    Array[File]? i1_fastq 
    String sample_id 
    String fastq_suffix 
    File whitelist 
    File r1_fastq 
    Int index 
    command {
    }
    output {
        File bam_output = ""
        Array[File]? non_optional_i1_fastq1 = []
        File? bam_output1 = ""
        File? bam_output2 = ""
        File barcoded_bam1 = ""
    }
}
workflow w {
    Array[File] r2_fastq 
    Array[File]? i1_fastq 
    String sample_id 
    String fastq_suffix 
    File whitelist 
    Array[File] r1_fastq 
    Array[Int] indices  = range(length(r1_fastq))
    scatter (index in indices) {
        call Optimus__scatter_index_body  {
            input:  r2_fastq=r2_fastq,
            i1_fastq=i1_fastq,
            sample_id=sample_id,
            fastq_suffix=fastq_suffix,
            whitelist=whitelist,
            r1_fastq=r1_fastq,
            index=index
        }
    }
    output {
        Array[Int] out_indices = indices
        Array[File] out_Optimus__scatter_index_body_bam_output = Optimus__scatter_index_body.bam_output
        Array[Array[File]?] out_Optimus__scatter_index_body_non_optional_i1_fastq1 = Optimus__scatter_index_body.non_optional_i1_fastq1
        Array[File?] out_Optimus__scatter_index_body_bam_output1 = Optimus__scatter_index_body.bam_output1
        Array[File?] out_Optimus__scatter_index_body_bam_output2 = Optimus__scatter_index_body.bam_output2
        Array[File] out_Optimus__scatter_index_body_barcoded_bam1 = Optimus__scatter_index_body.barcoded_bam1
    }
}
```

**Comment by orodeh** [comment]
I suspect that dxWDL is dropping the array access in `r1_fastq[index]` during one of the rewrite steps. Can you try modifying the original WDL to define it as a separate variable?

**Comment by orodeh** [comment]
@mckinsel, if you can create a reduced example, I'll work on it.

**Comment by mckinsel** [comment]
Okay, I'll see if I can come up with something.

**Comment by mckinsel** [comment]
Well, I'm having a hard time replicating it more simply. But this seems to fix it: https://github.com/HumanCellAtlas/skylab/commit/a8be87794de640f601a95709e719134a8d7adbf8

**Comment by orodeh** [comment]
It seems the problem is that the array access is dropped during workflow rewrite. That is helpful, thanks!

**Comment by orodeh** [comment]
The following situation confused the free variable analysis algorithm:
(1) a workflow calls a task
(2) the task and the workflow have an input with the same name but different types. 

For example, workflow `w` calls task `PTAsays`, both use input `fruit`, but it has types `Array[String]` and `String`.

```wdl
workflow w {
    Array[String] fruit = ["Banana", "Apple"]
    scatter (index in indices) {
        call PTAsays {
            input: fruit = fruit[index], y = " is good to eat"
        }
        call Add { input:  a = 2, b = 4 }
    }
}

task PTAsays {
    String fruit                                                                                                             
    String y                                                                                                                 
     ...                                                                                                                     
}
```

Fixed in version 0.69.
